### PR TITLE
feature: implement oldValue in jcache

### PR DESCRIPTION
**Pull request by testower**
Solves #3511

### CODE REVIEW

**Comment by mrniko** [comment]
Thanks for suggested pr. According to your changes old values are sent for any listener even if old value wasn't requested. This may increase pub/sub traffic for customers who don't need it.

**Comment by testower** [comment]
Thanks for the feedback, that makes sense. I'm happy to give it a shot. Let me know if you have any pointers or similar concepts already I can look at for guidance.

**Comment by mrniko** [comment]
You need to introduce oldvaluelistneres counter like with name generated like in getTimeoutSetName() method. Use `incr` command in `registerCacheEntryListener` method and `decr` in `deregisterCacheEntryListener` method. Don't send old value if counter value == false.

**Comment by testower** [comment]
Looks like your message crossed my draft implementation. I think we're on the same page. I have implemented for updated only, but will complete the other listeners if you think I'm on the right track here.

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 1]
```diff
@@ -157,6 +157,8 @@ String getRemovedChannelName() {
         return "jcache_removed_channel:{" + getName() + "}";
     }
 
+    String getUpdatedChannelOldValueRequiredCounter() { return "jcache_updated_old_value_required_counter:{" + getName() + "}"; }
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:160]: can we make it shorter like `jcache_old_value_listeners` ?

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 2]
```diff
@@ -400,16 +402,16 @@ private boolean putValueLocked(K key, Object value) {
                       + "elseif ARGV[2] ~= '-1' then "
                           + "redis.call('hset', KEYS[1], ARGV[4], ARGV[5]); "
                           + "redis.call('zadd', KEYS[2], ARGV[2], ARGV[4]); "
-                          + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5]); "                      
+                          + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:405]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 3]
```diff
@@ -400,16 +402,16 @@ private boolean putValueLocked(K key, Object value) {
...
                           + "redis.call('publish', KEYS[5], msg); "
-                          + "local syncMsg = struct.pack('Lc0Lc0d', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], ARGV[6]); "
+                          + "local syncMsg = struct.pack('Lc0Lc0Lc0d', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value), ARGV[6]); "
                           + "local syncs = redis.call('publish', KEYS[8], syncMsg); "
                           + "return {1, syncs};"
                       + "else "
                           + "redis.call('hset', KEYS[1], ARGV[4], ARGV[5]); "
-                          + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5]); "
+                          + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:412]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 4]
```diff
@@ -482,27 +484,27 @@ private boolean putValueLocked(K key, Object value) {
...
-                    + "local value = redis.call('hget', KEYS[1], ARGV[i]);"
                     + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[i]), ARGV[i], string.len(tostring(value)), tostring(value)); "
                     + "redis.call('publish', KEYS[4], msg); "
                     + "local syncMsg = struct.pack('Lc0Lc0d', string.len(ARGV[i]), ARGV[i], string.len(tostring(value)), tostring(value), ARGV[4]); "
                     + "syncs = syncs + redis.call('publish', KEYS[7], syncMsg); "
                 + "elseif ARGV[2] ~= '-1' then "
                     + "redis.call('hset', KEYS[1], ARGV[i], ARGV[i+1]); "
                     + "redis.call('zadd', KEYS[2], ARGV[2], ARGV[i]); "
-                    + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[i]), ARGV[i], string.len(ARGV[i+1]), ARGV[i+1]); "                      
+                    + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[i]), ARGV[i], string.len(ARGV[i+1]), ARGV[i+1], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:498]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 5]
```diff
@@ -482,27 +484,27 @@ private boolean putValueLocked(K key, Object value) {
...
                     + "redis.call('publish', KEYS[5], msg); "
-                    + "local syncMsg = struct.pack('Lc0Lc0d', string.len(ARGV[i]), ARGV[i], string.len(ARGV[i+1]), ARGV[i+1], ARGV[4]); "
+                    + "local syncMsg = struct.pack('Lc0Lc0Lc0d', string.len(ARGV[i]), ARGV[i], string.len(ARGV[i+1]), ARGV[i+1], string.len(tostring(value)), tostring(value), ARGV[4]); "
                     + "syncs = syncs + redis.call('publish', KEYS[8], syncMsg); "
                     + "added = added + 1;"
                 + "else "
                     + "redis.call('hset', KEYS[1], ARGV[i], ARGV[i+1]); "
-                    + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[i]), ARGV[i], string.len(ARGV[i+1]), ARGV[i+1]); "
+                    + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[i]), ARGV[i], string.len(ARGV[i+1]), ARGV[i+1], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:505]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 6]
```diff
@@ -1229,16 +1245,16 @@ public void put(K key, V value) {
                       + "elseif ARGV[2] ~= '-1' then " 
                           + "redis.call('hset', KEYS[1], ARGV[4], ARGV[5]); "
                           + "redis.call('zadd', KEYS[2], ARGV[2], ARGV[4]); "
-                          + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5]); "
+                          + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:1248]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 7]
```diff
@@ -1229,16 +1245,16 @@ public void put(K key, V value) {
...
                           + "redis.call('publish', KEYS[5], msg); "
-                          + "local syncMsg = struct.pack('Lc0Lc0d', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], ARGV[6]); "
+                          + "local syncMsg = struct.pack('Lc0Lc0Lc0d', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value), ARGV[6]); "
                           + "local syncs = redis.call('publish', KEYS[8], syncMsg); "
                           + "return {1, value, syncs};"
                       + "else " 
                           + "redis.call('hset', KEYS[1], ARGV[4], ARGV[5]); "
-                          + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5]); "
+                          + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:1255]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 8]
```diff
@@ -1302,16 +1318,16 @@ public void put(K key, V value) {
                   + "elseif ARGV[2] ~= '-1' then " 
                       + "redis.call('hset', KEYS[1], ARGV[4], ARGV[5]); "
                       + "redis.call('zadd', KEYS[2], ARGV[2], ARGV[4]); "
-                      + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5]); "
+                      + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[5]), ARGV[5], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:1321]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 9]
```diff
@@ -2208,20 +2224,20 @@ private long replaceValueLocked(K key, V oldValue, V newValue) {
...
-                           + "local value = redis.call('hget', KEYS[1], ARGV[4]); "
                            + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(tostring(value)), tostring(value)); "
                            + "redis.call('publish', KEYS[3], msg); "
                            + "local syncMsg = struct.pack('Lc0Lc0d', string.len(ARGV[4]), ARGV[4], string.len(tostring(value)), tostring(value), ARGV[7]); "
                            + "return redis.call('publish', KEYS[5], syncMsg); "
                        + "elseif ARGV[2] ~= '-1' then " 
                            + "redis.call('hset', KEYS[1], ARGV[4], ARGV[6]); "
                            + "redis.call('zadd', KEYS[2], ARGV[2], ARGV[4]); "
-                           + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[6]), ARGV[6]); "
+                           + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[6]), ARGV[6], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2238]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 10]
```diff
@@ -2296,11 +2312,11 @@ private long replaceValueLocked(K key, V oldValue, V newValue) {
                   + "elseif ARGV[2] ~= '-1' then " 
                       + "redis.call('hset', KEYS[1], ARGV[4], ARGV[6]); "
                       + "redis.call('zadd', KEYS[2], ARGV[2], ARGV[4]); "
-                      + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[6]), ARGV[6]); "
+                      + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[4]), ARGV[4], string.len(ARGV[6]), ARGV[6], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2315]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 11]
```diff
@@ -2404,26 +2420,26 @@ private boolean replaceValueLocked(K key, V value) {
...
-                  + "local value = redis.call('hget', KEYS[1], ARGV[3]); "
                   + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[3]), ARGV[3], string.len(tostring(value)), tostring(value)); "
                   + "redis.call('publish', KEYS[3], msg); "
                   + "local syncMsg = struct.pack('Lc0Lc0d', string.len(ARGV[3]), ARGV[3], string.len(tostring(value)), tostring(value), ARGV[5]); "
                   + "return redis.call('publish', KEYS[5], syncMsg); "
               + "elseif ARGV[1] ~= '-1' then "
                   + "redis.call('hset', KEYS[1], ARGV[3], ARGV[4]); "
                   + "redis.call('zadd', KEYS[2], ARGV[1], ARGV[3]); "
-                  + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[3]), ARGV[3], string.len(ARGV[4]), ARGV[4]); "
+                  + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[3]), ARGV[3], string.len(ARGV[4]), ARGV[4], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2434]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 12]
```diff
@@ -2462,11 +2478,11 @@ private boolean replaceValueLocked(K key, V value) {
...
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2481]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 13]
```diff
@@ -2497,11 +2513,11 @@ private boolean replaceValueLocked(K key, V value) {
               + "elseif ARGV[1] ~= '-1' then " 
                   + "redis.call('hset', KEYS[1], ARGV[3], ARGV[4]); "
                   + "redis.call('zadd', KEYS[2], ARGV[1], ARGV[3]); "
-                  + "local msg = struct.pack('Lc0Lc0', string.len(ARGV[3]), ARGV[3], string.len(ARGV[4]), ARGV[4]); "
+                  + "local msg = struct.pack('Lc0Lc0Lc0', string.len(ARGV[3]), ARGV[3], string.len(ARGV[4]), ARGV[4], string.len(tostring(value)), tostring(value)); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2516]: `oldValueRequired` should be checked

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 14]
```diff
@@ -2540,15 +2556,15 @@ private V getAndReplaceValueLocked(K key, V value) {
...
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2559]: `oldValueRequired` should be checked

**Review by mrniko — CHANGES_REQUESTED** [review]

**Comment by mrniko** [comment]
I added few comments. Implementation looks right

**Comment by testower** [comment]
I think there is a problem in the listener logic. When the first listener registers without requiring the old value it will instantiate a codec that doesn't expect it, and then another listener registers requiring the old value, then the first listener will start receiving it as well, and its codec will need to be instantiated as well. This will only result in an error when sync methods are used I think. I will try to produce a test case and look into how to solve it. Let me know if you have any suggestions.

**Comment by mrniko** [comment]
I think codec needs to be modified and data with 0 length should be sent when we don't need old value.

**Comment by testower** [comment]
I have updated the PR with a suggested solution. In the end, a signed short value of `-1` indicates that there is no old value in the message and the decoder can skip trying to decode it. Let me know if that's sort of what you had in mind.

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 15]
```diff
@@ -2403,31 +2496,45 @@ private boolean replaceValueLocked(K key, V value) {
         if (containsKey(key)) {
             double syncId = ThreadLocalRandom.current().nextDouble();
             Long updateTimeout = getUpdateTimeout();
-        Long syncs = evalWrite(getRawName(), codec, RedisCommands.EVAL_LONG,
-                "if ARGV[1] == '0' then "
+            Long syncs = evalWrite(getRawName(), codec, RedisCommands.EVAL_LONG,
+                "local value = redis.call('hget', KEYS[1], ARGV[3]); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2500]: call `hget` only if it's really needed.

File: redisson/src/main/java/org/redisson/jcache/JCache.java [hunk 16]
```diff
@@ -2529,8 +2660,8 @@ private V getAndReplaceValueLocked(K key, V value) {
             Long updateTimeout = getUpdateTimeout();
             double syncId = ThreadLocalRandom.current().nextDouble();
             Long syncs = evalWrite(getRawName(), codec, RedisCommands.EVAL_LONG,
-                "if ARGV[1] == '0' then "
-                  + "local value = redis.call('hget', KEYS[1], ARGV[3]); "
+                    "local value = redis.call('hget', KEYS[1], ARGV[3]); "
```
Comment (mrniko) [inline redisson/src/main/java/org/redisson/jcache/JCache.java:2663]: call `hget` only if it's really needed.

**Review by mrniko — CHANGES_REQUESTED** [review]

**Comment by mrniko** [comment]
Thanks much for this contribution!

**Pull request merged**